### PR TITLE
P2446R2: `views::as_rvalue`

### LINKED ISSUE #2929: P2446R2 `views::as_rvalue`
[P2446R2](https://wg21.link/P2446R2) `views::as_rvalue`

Feature-test macro (expected):
```cpp
#define __cpp_lib_ranges_as_rvalue 202207L
```

### PR DESCRIPTION
**Pull request by JMazurkiewicz**
Implements [P2446R2](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2446r2.html) - `views::as_rvalue`. Closes #2929.

Notes:
* ~~Marked as draft - I'm unable to implement full test coverage due to iterator unwrapping problems.~~ Not anymore - I've implemented custom `test_equal` function that should be replaced with `ranges::equal` after #3009 is fixed,
* Just like in #2981: 
  > I've `std::`-qualified all calls to `std::forward` and `std::move` in tests, because next Clang release (15) is going to [warn about "unqualified std cast calls"](https://clang.llvm.org/docs/DiagnosticsReference.html#wunqualified-std-cast-call).

### CODE REVIEW

File: stl/inc/ranges
Comment (strega-nil-ms) [inline stl/inc/ranges:1632]:
```suggestion
    template <view _Vw>
        requires input_range<_Vw>
```

consistency with the standard

File: stl/inc/ranges
Comment (strega-nil-ms) [inline stl/inc/ranges:1688]:
```suggestion
        _NODISCARD constexpr auto size() noexcept(noexcept(_Fake_decay_copy(_RANGES size(_Range)))) // strengthened
```

we decay-copy the return value of `_RANGES size()`. Same below.

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]
Minor changes requested, otherwise LGTM!

File: stl/inc/ranges
Comment (strega-nil-ms) [reply stl/inc/ranges:1688]: @CaseyCarter please confirm? We don't do this elsewhere and I think it's a bug.

File: stl/inc/ranges
Comment (JMazurkiewicz) [reply stl/inc/ranges:1632]: Here standard is inconsistent with itself, [`[range.as.rvalue.view]`](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2446r2.html#class-template-as_rvalue_view-range.as.rvalue.view) says:
```c++
namespace std::ranges {
  template<input_range V>
    requires view<V>
  class as_rvalue_view : public view_interface<as_rvalue_view<V>>
  {
    V base_ = V(); // exposition only
//and so on
```

Comment (strega-nil-ms) [reply stl/inc/ranges:1632]: GFDI, fair enough.

File: stl/inc/ranges
Comment (CaseyCarter) [reply stl/inc/ranges:1688]: `ranges::size` always returns a prvalue.

Comment (StephanTLavavej) [reply stl/inc/ranges:1688]: And (as we discussed at the maintainer meeting) integer-like types don't throw exceptions - so only the actual call to `ranges::size` can throw.

**Review by strega-nil-ms — APPROVED** [review]
Thanks!

File: tests/std/tests/P2446R2_views_as_rvalue/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/P2446R2_views_as_rvalue/test.cpp:207]: @CaseyCarter noted that we should verify that `begin()` and `end()` return `move_iterator` and/or `move_sentinel`; the test currently doesn't mention these types directly.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks, this looks great! I pushed minor changes (FYI @strega-nil-ms after you approved) for nitpicky style issues and a slight expansion of test coverage.

File: tests/std/tests/P2446R2_views_as_rvalue/test.cpp
Comment (JMazurkiewicz) [reply tests/std/tests/P2446R2_views_as_rvalue/test.cpp:360]: Whoops, I thought it was included by `<ranges>`...

File: tests/std/tests/P2446R2_views_as_rvalue/test.cpp
Comment (JMazurkiewicz) [reply tests/std/tests/P2446R2_views_as_rvalue/test.cpp:416]: ...but this one, `<iterator>`, is included by `<ranges>`: http://eel.is/c++draft/ranges.syn

Comment (CaseyCarter) [reply tests/std/tests/P2446R2_views_as_rvalue/test.cpp:416]: Yes: we often take advantage of the `<ranges>` => `<iterator>` => `<concepts>` chain in Ranges tests. That said, it's not worth the bother to remove now that Stephan has added `<iterator>`.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks again for implementing this C++23 feature and helping users write more elegant, more efficient code! :heart_eyes_cat: :rocket: :tada:

**Pull request merged**
